### PR TITLE
Bug 1245247 - Reconciled login loading code into single method and added loading view to better manage state

### PR DESCRIPTION
**Pull request by sleroux**

### CODE REVIEW

**Comment by fluffyemily** [comment]
Only thing I would say is, if we are doing an account sync and the current Login view is empty (i.e. first time we run the app and connect to sync) then this VC just tells me that I have no logins, even though it's possible that I do have logins but they just haven't synced over yet. Is there any way that in that circumstance we can continue showing the loading view until the sync is complete when we either show the synced logins or the no logins view? For a while during testing I thought that password sync was broken because it took so long for my passwords to appear on first start.

**Comment by sleroux** [comment]
One problem I've seen is that syncing everything takes a long time but the logins come down pretty quickly. For example, if you were to start syncing, navigate to the logins list, go back then forth, there's a good chance that the logins will already be there. I was stumped as to why the logins took so long to reload after syncing but realized that we only reload the table view when the entire syncing process is finished. Ideally we would reload when login sync is complete which will dramatically improve the time it takes for logins to appear. 

This doesn't really solve the problem you mentioned though where it might look like a bug to the user. I think an additional UI hint informing the user that we are still syncing might alleviate that concern - like a banner over the status bar or something app-wide to let the user remember that we are in fact still syncing in order to set that expectation that your stuff might not be available yet. I like the idea of showing the no logins screen when the user is syncing because you actually don't have any logins yet.

**Comment by sleroux** [comment]
@rnewman Would it be difficult to add start/finish notifications for the each sync engine? I know we leverage the spinLock didSet observer to tell us when a full sync occurs but I'm not sure of the equivalent state change for each synchronizer.

**Comment by rnewman** [comment]
Easy enough to do. But I suspect what you're looking for is a "did change" observer, right?

**Comment by sleroux** [comment]
@rnewman What would the 'change' be? The change in the login synchronizer state?

**Comment by rnewman** [comment]
Logins themselves. You want to find out when logins have been added or removed, right? Not all login syncs do so — they might make no changes, or they might only sync data up, not down — and it's not only syncing (in theory) that could change the set of logins.

In this case, you'd show "No logins", then as soon as we're done bulk-inserting the downloaded ones from Sync, we'd refresh the list.

**Comment by sleroux** [comment]
True. What about using a DB trigger on the loginsL table so that way we get notified on updates/inserts?

**Comment by rnewman** [comment]
Sync would write into `loginsM`, but speaking more generally — triggers are pretty granular, and we don't do all of Sync's work in a single transaction, so you'd get `N` change notifications during a sync.

Also you'd need to register a (C) callback function into SQL to use triggers, which is a bit of a pain. Much easier to just have `SQLiteLogins` expose a little change tracking/notification thing.

**Comment by sleroux** [comment]
@rnewman @fluffyemily I added an observer on the existing login change event to the list view. I also attached the login change event to the various sync methods that get called when logins are inserted/updated so now the logins should reload when the logins have been added from sync.

**Comment by sleroux** [comment]
@rnewman @fluffyemily I added an observer on the existing login change event to the list view. I also attached the login change event to the various sync methods that get called when logins are inserted/updated so now the logins should reload when the logins have been added from sync.

File: Storage/SQL/SQLiteLogins.swift [hunk 1]
```diff
@@ -651,7 +651,7 @@ extension SQLiteLogins: SyncableLogins {
             ", sync_status = \(SyncStatus.Changed.rawValue) " +
         " WHERE guid = ?"
 
-        return self.db.run(update, withArgs: args)
+        return self.db.run(update, withArgs: args) >>> effect(self.notifyLoginDidChange)
```
Comment (rnewman) [inline Storage/SQL/SQLiteLogins.swift]: These lines will be called at least once for every login that changes during a sync, which is probably not what you want. Instead, call this *once* at the end of `LoginsSynchronizer.synchronizeLocalLogins`.

**Comment by sleroux** [comment]
@rnewman Good call. I've updated the patch to only notify from the synchronizer:

https://github.com/mozilla/firefox-ios/commit/6f4af964274eddfc1b6db48019f9f817cd70db58

File: Sync/Synchronizers/LoginsSynchronizer.swift [hunk 2]
```diff
@@ -162,7 +162,9 @@ public class LoginsSynchronizer: IndependentRecordSynchronizer, Synchronizer {
             let lm = response.metadata.lastModifiedMilliseconds!
             log.debug("Applying incoming password records from response timestamped \(ts), last modified \(lm).")
             log.debug("Records header hint: \(response.metadata.records)")
-            return self.applyIncomingToStorage(logins, records: response.value, fetched: lm)
+            return self.applyIncomingToStorage(logins, records: response.value, fetched: lm) >>> effect {
+                NSNotificationCenter.defaultCenter().postNotificationName(NotificationDataLoginDidChange, object: nil)
```
Comment (rnewman) [inline Sync/Synchronizers/LoginsSynchronizer.swift]: Don't use `NotificationDataLoginDidChange`! We listen for that to trigger a sync:

https://github.com/mozilla/firefox-ios/search?utf8=%E2%9C%93&q=onLoginDidChange

Comment (rnewman) [inline Sync/Synchronizers/LoginsSynchronizer.swift]: If you do this, you'll get a second sync immediately after the first, and if that adds records you'll get a third… and so on.

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 3]
```diff
@@ -78,7 +80,8 @@ class LoginListViewController: UIViewController {
         super.viewDidLoad()
 
         let notificationCenter = NSNotificationCenter.defaultCenter()
-        notificationCenter.addObserver(self, selector: Selector("SELonProfileDidFinishSyncing"), name: NotificationProfileDidFinishSyncing, object: nil)
+        notificationCenter.addObserver(self, selector: Selector("SELreloadLogins"), name: NotificationProfileDidFinishSyncing, object: nil)
+        notificationCenter.addObserver(self, selector: Selector("SELreloadLogins"), name: NotificationDataLoginDidChange, object: nil)
```
Comment (rnewman) [inline Client/Frontend/Login Management/LoginListViewController.swift]: I don't think you need to listen to both. I think you need one and one only, which we might call `NotificationDataRemoteLoginChangesWereApplied`, and we issue that in the place you currently send `LoginDidChange`.

File: Client/Frontend/Login Management/LoginListViewController.swift [hunk 4]
```diff
@@ -160,26 +168,28 @@ class LoginListViewController: UIViewController {
         }
     }
 
-    deinit {
-        let notificationCenter = NSNotificationCenter.defaultCenter()
-        notificationCenter.removeObserver(self, name: NotificationProfileDidFinishSyncing, object: nil)
+    private func loadLogins(query: String? = nil) -> Success {
+        loadingStateView.hidden = false
+        activeLoginQuery = profile.logins.searchLoginsWithQuery(query).bindQueue(dispatch_get_main_queue(), f: reloadTableWithResult)
+        return activeLoginQuery!
```
Comment (rnewman) [inline Client/Frontend/Login Management/LoginListViewController.swift]: I don't understand why this needs `!`. `bindQueue` doesn't return an optional.

**Comment by sleroux** [comment]
@rnewman Updated with comment fixes

**Comment by sleroux** [comment]
I'm going to land this after I rebase with the test fixes from https://github.com/mozilla/firefox-ios/pull/1515. Don't feel comfortable landing before I know all these tests pass.

**Pull request merged**